pkg/scm/git/clone.go

```diff
@@ -19,86 +22,94 @@ type Clone struct {
 func (c *Clone) Download(config *api.Config) (*api.SourceInfo, error) {
 	targetSourceDir := filepath.Join(config.WorkingDir, api.Source)
 	config.WorkingSourceDir = targetSourceDir
-	var info *api.SourceInfo
-	hasRef := len(config.Ref) > 0
-	cloneConfig := api.CloneConfig{Quiet: true, Recursive: !config.IgnoreSubmodules && !hasRef}
 
 	ok, err := c.ValidCloneSpec(config.Source)
 	if err != nil {
 		return nil, err
 	}
+	if !ok {
+		glog.Errorf("Clone.Download was passed an invalid source %s", config.Source)
+		return nil, fmt.Errorf("invalid source %s", config.Source)
+	}
 
-	if ok {
-		if len(config.ContextDir) > 0 {
-			targetSourceDir = filepath.Join(config.WorkingDir, api.ContextTmp)
-			glog.V(1).Infof("Downloading %q (%q) ...", config.Source, config.ContextDir)
-		} else {
-			glog.V(1).Infof("Downloading %q ...", config.Source)
-		}
+	ref := "HEAD"
+	if config.Ref != "" {
+		ref = config.Ref
+	}
 
-		// If we have a specific checkout ref, use submodule update instead of recursive
-		// Otherwise the versions will be incorrect.
-		if hasRef && !config.IgnoreSubmodules {
-			glog.V(2).Infof("Cloning sources (deferring submodule init) into %q", targetSourceDir)
-		} else if !hasRef && !config.IgnoreSubmodules {
-			glog.V(2).Infof("Cloning sources and all Git submodules (recursive) into %q", targetSourceDir)
-		} else {
-			glog.V(2).Infof("Cloning sources (ignoring submodules) into %q", targetSourceDir)
-		}
+	if strings.HasPrefix(config.Source, "file://") {
+		s := strings.TrimPrefix(config.Source, "file://")
 
-		if err := c.Clone(config.Source, targetSourceDir, cloneConfig); err != nil {
-			glog.V(1).Infof("Git clone failed: %+v", err)
-			return nil, err
-		}
-
-		if hasRef {
-			if err := c.Checkout(targetSourceDir, config.Ref); err != nil {
-				return nil, err
-			}
-			glog.V(1).Infof("Checked out %q", config.Ref)
-			if !config.IgnoreSubmodules {
-				if err := c.SubmoduleUpdate(targetSourceDir, true, true); err != nil {
-					return nil, err
-				}
-				glog.V(1).Infof("Updated submodules for %q", config.Ref)
-			}
-		}
-
-		if len(config.ContextDir) > 0 {
-			originalTargetDir := filepath.Join(config.WorkingDir, api.Source)
-			c.RemoveDirectory(originalTargetDir)
-			// we want to copy entire dir contents, thus we need to use dir/. construct
-			path := filepath.Join(targetSourceDir, config.ContextDir) + string(filepath.Separator) + "."
-			err := c.Copy(path, originalTargetDir)
+		if util.UsingCygwinGit {
+			var err error
+			s, err = util.ToSlashCygwin(s)
 			if err != nil {
+				glog.V(1).Infof("ToSlashCygwin failed: %+v", err)
```

Comment:
Why do we have the `+` here? If the error has an underlying struct type, it will print a struct with field names instead of the error message -- I think that's not what we want in log level 1?